### PR TITLE
Key completion for dataset access

### PR DESCRIPTION
**Pull request by ElenaKhaustova**
## Description
Solves https://github.com/kedro-org/kedro/issues/3914

## Development notes

- Implemented `IPython` and `Jupyter` key completions for accessing datasets in the catalog. The solution is based on defining `_ipython_key_completions_()` for `_FrozenDatasets` class.

Screenshot 2024-07-01 at 08 55 01

- Original names (with non-word characters) are suggested as keys but those that are replaced with `__` (as for properties) are accepted as well.

Screenshot 2024-07-01 at 08 53 32

- Manually tested with transcoded datasets and configuration patterns.

## Developer Certificate of Origin
We need all contributions to comply with the [Developer Certificate of Origin (DCO)](https://developercertificate.org/). All commits must be signed off by including a `Signed-off-by` line in the commit message. [See our wiki for guidance](https://github.com/kedro-org/kedro/wiki/Guidelines-for-contributing-developers/).

If your PR is blocked due to unsigned commits, then you must follow the instructions under "Rebase the branch" on the GitHub Checks page for your PR. This will retroactively add the sign-off to all unsigned commits and allow the DCO check to pass.

## Checklist

- [ ] Read the [contributing](https://github.com/kedro-org/kedro/blob/main/CONTRIBUTING.md) guidelines
- [ ] Signed off each commit with a [Developer Certificate of Origin (DCO)](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/managing-repository-settings/managing-the-commit-signoff-policy-for-your-repository)
- [ ] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [ ] Updated the documentation to reflect the code changes
- [ ] Added a description of this change in the [`RELEASE.md`](https://github.com/kedro-org/kedro/blob/main/RELEASE.md) file
- [ ] Added tests to cover my changes
- [ ] Checked if this change will affect Kedro-Viz, and if so, communicated that with the Viz team

### CODE REVIEW

File: kedro/io/data_catalog.py [hunk 1]
```diff
@@ -105,28 +105,36 @@ def __init__(
...
+            return
         msg = "Operation not allowed! "
         if key in self.__dict__:
             msg += "Please change datasets through configuration."
         else:
             msg += "Please use DataCatalog.add() instead."
         raise AttributeError(msg)
 
+    def _ipython_key_completions_(self) -> list[str]:
+        return list(self._original_names)
```
Comment (datajoely) [inline kedro/io/data_catalog.py:133]: So there is an argument that simplicity may be better, but what about including the `dataset.__class__.__name__` so you would see something like `shuttles (CSVDataset)`

Comment (datajoely) [reply kedro/io/data_catalog.py:133]: I would also artificially place `parameters` at the very bottom to discourage its use

Comment (ElenaKhaustova) [reply kedro/io/data_catalog.py:133]: > So there is an argument that simplicity may be better, but what about including the `dataset.__class__.__name__` so you would see something like `shuttles (CSVDataset)`?

I think it's a good idea and I see the value. The only concern is whether we are fine with those keys. Cause with tab completion we will get `catalog.datasets["shuttles (CSVDataset)"]` which will work but not sure it looks good from the syntax side.

Comment (ElenaKhaustova) [reply kedro/io/data_catalog.py:133]: I'll check if it's possible to provide some metadata which is not part of the key but still displayed with `IPython` and `Jupyter` UI.

File: kedro/io/data_catalog.py [hunk 2]
```diff
@@ -105,28 +105,36 @@ def __init__(
...
+
+    def __getitem__(self, key: str) -> Any:
+        return self.__dict__[_sub_nonword_chars(key)]
```
Comment (noklam) [inline kedro/io/data_catalog.py:136]: Can we change `self._original_names` to a dictionary and modify the `__getitem__` to access data through this mapping?

i.e. `self.__dict__[self._original_names[key]]`

**Review by noklam — COMMENTED** [review]
Generally looks good, I have some questions:
image

1. When I try to prompt the autocompletion, it seems to prompt me to use the single quote instead of the more common `"` in Python. Is this something we can control or it's how IPython completion works? If I type `catalog.datasets["]` and hit tab then it would finish the autocompletion without the finishing `"`, which means a few more keystrokes.  This is not a deal breaker but just want to check if we can make this experience smoother.

File: kedro/io/data_catalog.py (hunk 2 above)
Comment (ElenaKhaustova) [reply kedro/io/data_catalog.py:136]: I aimed to make both original names (with non-word characters) - `catalog.datasets["params:model_option"]` and transformed names - `catalog.datasets["params__model_option"]` work, so ended up with the current solution. So there's less discrepancy with by attribute access `catalog.datasets.params__model_option`

**Comment by ElenaKhaustova** [comment]
@datajoely, regarding your question about placing parameters at the bottom - sorting is applied internally in `ipython` `completer.py` after `_ipython_key_completions_()` is called, so I cannot force a custom order. Keys are always in lexicographic order.

**Comment by ElenaKhaustova** [comment]
> Generally looks good, I have some questions: image
> 
> 1. When I try to prompt the autocompletion, it seems to prompt me to use the single quote instead of the more common `"` in Python. Is this something we can control or it's how IPython completion works? If I type `catalog.datasets["]` and hit tab then it would finish the autocompletion without the finishing `"`, which means a few more keystrokes.  This is not a deal breaker but just want to check if we can make this experience smoother.

Yesk, that's the way [IPython completion](https://github.com/ipython/ipython/blob/a499dbd507a92ea0087eca0f53e550fc838f0580/IPython/core/completer.py#L2489) works and there's no any API to customise it from the outside. I see [some logic](https://github.com/ipython/ipython/blob/a499dbd507a92ea0087eca0f53e550fc838f0580/IPython/core/completer.py#L2558) for matching the closing quote but it's not straightforward why it's not working. So there's not much that we can do but create an issue and ask.

**Review by noklam — APPROVED** [review]

**Comment by noklam** [comment]
Thanks for checking, I think this is good enough.

**Comment by astrojuanlu** [comment]
> Sounds good! @ElenaKhaustova , I petition that we drop the `parameters` god object in `1.0.0` - cc @astrojuanlu

Not sure if related, but https://github.com/kedro-org/kedro/pull/3732#discussion_r1537828647

> Shouldn't we redesign the `DataCatalog` API instead so that parameters are first class citizens, and not fake datasets?

The recent research didn't surface any user problems around that, but if we're to change anything in that regard, probably the time to look at it is now. Shall we open a separate issue about it?

**Comment by merelcht** [comment]
We still have the issue #2240 originally intended for a big research piece around parameter usage.

**Comment by ElenaKhaustova** [comment]
> > Sounds good! @ElenaKhaustova , I petition that we drop the `parameters` god object in `1.0.0` - cc @astrojuanlu
> 
> Not sure if related, but [#3732 (comment)](https://github.com/kedro-org/kedro/pull/3732#discussion_r1537828647)
> 
> > Shouldn't we redesign the `DataCatalog` API instead so that parameters are first class citizens, and not fake datasets?
> 
> The recent research didn't surface any user problems around that, but if we're to change anything in that regard, probably the time to look at it is now. Shall we open a separate issue about it?

The idea sounds good, at least from the performance point of view mentioned in https://github.com/kedro-org/kedro/pull/3732#discussion_r1537828647. I think it also falls into the discussion about making datasets of first-class citizens, which we'll have anyway in terms of future `DataCatalog` redesign work. Vote for opening an issue.

**Comment by astrojuanlu** [comment]
Left a comment at https://github.com/kedro-org/kedro/issues/3893#issuecomment-2203406999 and marking the off-topic comments as such.

**Review by astrojuanlu — APPROVED** [review]
Works like a charm! ✨ 

image

**Pull request merged**